### ISSUE TITLE
twisted.web.distrib fails for resources which write strings longer than the banana length limit to the request

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3212]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3212|
|Type|defect|
|Created|2008-04-28 15:12:28Z|
|Branch|https://github.com/twisted/twisted/tree/distrib-large-write-3212|

From the distributed side, resources are given a request object which maps to some PB calls.  However, `IRequest.write` imposes no limit on the length of the string passed to it.  The distrib implementation does impose a limit though, due to banana's length limit.

Here's a demonstration:

```python
from os import remove
from sys import stdout

from twisted.python.log import startLogging
from twisted.web.distrib import ResourcePublisher, ResourceSubscription
from twisted.web.resource import Resource
from twisted.web.server import Site
from twisted.internet import reactor
from twisted.spread.pb import PBServerFactory

class FatResource(Resource):
    def render(self, request):
        return "x" * 1024 * 1024


def main():
    startLogging(stdout)
    try:
        remove("distrib-demo")
    except OSError:
        pass
    reactor.listenUNIX("distrib-demo", PBServerFactory(ResourcePublisher(Site(FatResource()))))
    root = Resource()
    root.putChild("", ResourceSubscription("unix", "distrib-demo"))
    reactor.listenTCP(8080, Site(root))
    reactor.run()

if __name__ == '__main__':
    main()
```

The `IRequest.write` implementation must be able to handle strings of any length, so it will probably need to do some buffering of some sort.  Performance might be better if producer/consumer got involved somehow, but maybe that's a separate optimization task for later.

<details><summary>Searchable metadata</summary>

```
trac-id__3212 3212
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_distrib_large_write_3212 branches/distrib-large-write-3212
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1209395548000000 1209395548000000
changetime__1266073175000000 1266073175000000
version__None None
owner__ 
cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28185]) Branching to 'distrib-large-write-3212'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Fixed in a rather lame way in the branch.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|spiv set owner to @exarkun|
|-|-|

The basic approach and tests seem fine to me.  I agree that maybe it could be done better, but this is a good incremental improvement.

This bit looks odd:

```
-        return res.render(request)
+        result = res.render(request)
+        if result is not server.NOT_DONE_YET:
+            request.write(result)
+            request.finish()
+        return server.NOT_DONE_YET
```

It appears a bit weird to return `NOT_DONE_YET` after calling `request.finish()`.  I suspect this is fine, and probably necessary if not returning a string, but it certainly looks odd.

Other comments:

 * why not use the `SIZE_LIMIT` constant in the implementation of `write`?
 * are there any compatibility concerns here if one side of a `t.w.distrib` deployment uses this version of Twisted, but the other uses an old one?  i.e. will the possibility of multiple remote calls to `write` rather than just one cause problems?  Off the top of my head I don't think so, but I'd appreciate you saying that you think it's fine too... either way, I guess it'll be no worse than the existing behaviour.

But basically, if tests pass, I'd be happy to see this merged as-is.  My comments are all suggestions, not required changes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `assigned`|
|-|-|

> It appears a bit weird to return NOT_DONE_YET after calling request.finish().

I agree.  I always try to let it return `None` in cases like this and then get yelled at for `None` not being a string.  It would be really great to have proper producer/consumer support in these APIs. :)

> why not use the SIZE_LIMIT constant in the implementation of write? 

Good question.  No idea.  I changed it to do so in r28193.

> are there any compatibility concerns here

Nope, a new version of the client code will correctly talk to an old version of the server code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [28194]) Merge distrib-large-write-3212

Author: exarkun
Reviewer: spiv
Fixes: [#3212](https://github.com/twisted/twisted/issues/3212)

In Twisted Web distrib nodes, avoid writing strings that are longer than
allowed by Banana to the request by automatically splitting up long strings
into multiple smaller pieces.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
